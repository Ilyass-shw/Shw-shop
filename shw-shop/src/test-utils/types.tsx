import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';

export type StoreType = ReturnType<typeof configureStore>;
export type RenderReturn = ReturnType<typeof render>;
export type SliceParameters = Parameters<typeof createSlice>;
export type UiType = React.ReactElement;

export interface WrapperParams {
  MyStore: StoreType;
  WithRouter?: boolean;
}

export interface useFormProps {
  email: string;
  phone: number;
  buyerAcceptMarketing: string;
}
