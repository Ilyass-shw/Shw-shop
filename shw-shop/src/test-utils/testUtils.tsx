import React, { ReactElement } from 'react';
import { render as rtlRender, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import store, { makeCustomTestStore } from './faksestore';
import { RenderReturn, StoreType, useFormProps } from './types';
import { BrowserRouter } from 'react-router-dom';
import { realStoreType } from '../App/store';
import { useForm, FormProvider } from 'react-hook-form';

export const render = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(ui);
};

// ============= ============= ============= ============= =============

export const renderWithStore = (
  ui: React.ReactElement,
  MyStore: realStoreType,
  WithRouter?: 'withRouter',
): RenderReturn => {
  if (WithRouter) {
    return rtlRender(
      <Provider store={MyStore}>
        <BrowserRouter>{ui}</BrowserRouter>
      </Provider>,
    );
  } else {
    return rtlRender(<Provider store={MyStore}>{ui}</Provider>);
  }
};

// ============= ============= ============= ============= =============

export const renderWithRouter = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(<BrowserRouter>{ui}</BrowserRouter>);
};

// ============= ============= ============= ============= =============

export const makeTestStore = (
  Name?: string,
  initialState?: Record<string, unknown>,
  reducers = {},
): StoreType => {
  if (Name && initialState) {
    const CustomTestStore = makeCustomTestStore(Name, initialState, reducers);
    const origDispatch = CustomTestStore.dispatch;
    CustomTestStore.dispatch = jest.fn(origDispatch);
    return CustomTestStore;
  } else {
    const origDispatch = store.dispatch as jest.Mock;
    store.dispatch = jest.fn(origDispatch);
    return store;
  }
};

// re-export everything
export * from '@testing-library/react';

// ============= ============= ============= ============= =============

export const renderWithFormProvider = (
  ui: ReactElement,
  WithRouter?: 'withRouter',
  { defaultValues = {} } = {},
): RenderResult => {
  // let reactHookFormMethods = {};

  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  };
  const WithRouterWrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return (
      <BrowserRouter>
        <FormProvider {...methods}>{children}</FormProvider>
      </BrowserRouter>
    );
  };

  if (!WithRouter) {
    return rtlRender(ui, { wrapper: Wrapper });
  }
  return rtlRender(ui, { wrapper: WithRouterWrapper });
};

// ============= ============= ============= ============= =============

export const renderWithReactHookForm = (
  ui: JSX.Element,
  saveData: jest.Mock,
): RenderResult => {
  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm<useFormProps>();
    return (
      <FormProvider {...methods}>
        <form
          onSubmit={methods.handleSubmit((data) => {
            saveData(data);
          })}
        >
          <BrowserRouter>{children}</BrowserRouter>
          <button type="submit" data-testid="testSubmitButton"></button>
        </form>
      </FormProvider>
    );
  };
  return rtlRender(<Wrapper>{ui}</Wrapper>);
};
