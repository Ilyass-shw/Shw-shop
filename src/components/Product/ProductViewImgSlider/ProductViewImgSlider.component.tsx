import styled from 'styled-components';

export const ImgContainer = styled.div`
  flex-shrink: 0;
  width: 90vw;
  align-self: center;
  @media screen and (min-width: 916px) {
    width: 60vw;
    display: flex;
    flex-direction: row-reverse;
  }
`;

// ============= ============= ============= ============= =============

export const AllImgs = styled.div`
  padding: 1rem 1rem;
  height: fit-content;
  width: fit-content;
  margin: auto;
  @media screen and (min-width: 916px) {
    padding: 0;
    margin: 0;
    align-self: center;
    display: flex;
    flex-direction: column;
    align-items: center;
  }
`;

// ============= ============= ============= ============= =============

type ImgBorderProps = { selected: boolean };

export const ImgBorder = styled.div<ImgBorderProps>`
  display: flex;
  margin: 0 0.5rem;
  align-items: flex-start;
  border: ${({ selected }) => (selected ? '2.5px solid black' : 'none')};
`;

// ============= ============= ============= ============= =============

export const Img = styled.img`
  width: 11vw;
  margin: 0.5rem;
  @media screen and (min-width: 550px) {
    width: 8vw;
  }
`;
