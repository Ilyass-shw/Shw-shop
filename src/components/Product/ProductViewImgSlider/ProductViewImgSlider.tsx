import React, { useState } from 'react';
import 'react-medium-image-zoom/dist/styles.css';
import { CustomButton } from '../../../App/App.component';
import ProductViewMainImg from '../ProductViewMainImg/ProductViewMainImg';

import {
  ImgContainer,
  AllImgs,
  ImgBorder,
  Img,
} from './ProductViewImgSlider.component';

interface productViewImgSliderProps {
  imgs: string[];
}
const ProductViewImgSlider: React.FC<productViewImgSliderProps> = ({
  imgs,
}) => {
  const [main, setMain] = useState(0);

  return (
    <ImgContainer>
      <ProductViewMainImg main={main} setMain={setMain} imgs={imgs} />
      <AllImgs data-testid='miniImgs'>
        {imgs.map((src, index) => (
          <CustomButton key={src} onClick={() => setMain(index)}>
            <ImgBorder selected={src === imgs[main]}>
              <Img loading="lazy" src={src} data-testid={'miniImg ' + src} />
            </ImgBorder>
          </CustomButton>
        ))}
      </AllImgs>
    </ImgContainer>
  );
};

export default ProductViewImgSlider;
