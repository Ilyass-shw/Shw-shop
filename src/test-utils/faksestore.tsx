import { configureStore } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { StoreType } from './types';

const fakeSlice = createSlice({
  name: 'name',
  initialState: {},
  reducers: {},
});

export default configureStore({
  reducer: {
    posts: fakeSlice.reducer,
  },
});

export const makeCustomTestStore = (
  Name: string,
  initialState: Record<string, unknown>,
  reducers = {},
): StoreType => {
  const FakeSlice = createSlice({
    name: Name,
    initialState,
    reducers,
  });

  return configureStore({
    reducer: {
      posts: FakeSlice.reducer,
    },
  });
};
