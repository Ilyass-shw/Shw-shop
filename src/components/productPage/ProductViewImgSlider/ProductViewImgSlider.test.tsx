import React from 'react';
import userEvent from '@testing-library/user-event';
import { renderWithRouter, screen } from '../../../test-utils/testUtils';
import ProductViewImgSlider from './ProductViewImgSlider';

describe('ProductViewImgSlider', () => {
  const imgs = ['src1', 'src2', 'src3'];

  it('should render', () => {
    renderWithRouter(<ProductViewImgSlider imgs={imgs} />);
  });

  it('should match Snapshot', () => {
    const { asFragment } = renderWithRouter(
      <ProductViewImgSlider imgs={imgs} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should change and display the right main img using the right arrow', () => {
    renderWithRouter(<ProductViewImgSlider imgs={imgs} />);
    const mainImg = screen.getByTestId('ZoomableImg') as HTMLImageElement;
    const rightArrow = screen.getByTestId('rightArrow');

    expect(mainImg.src).toContain('src1');
    userEvent.click(rightArrow);
    expect(mainImg.src).toContain('src2');
    userEvent.click(rightArrow);
    expect(mainImg.src).toContain('src3');
    userEvent.click(rightArrow);
    expect(mainImg.src).toContain('src1');
  });

  it('should render change and display the right main img using the left arrow', () => {
    renderWithRouter(<ProductViewImgSlider imgs={imgs} />);
    const mainImg = screen.getByTestId('ZoomableImg') as HTMLImageElement;
    const leftArrow = screen.getByTestId('leftArrow');

    expect(mainImg.src).toContain('src1');
    userEvent.click(leftArrow);
    expect(mainImg.src).toContain('src3');
    userEvent.click(leftArrow);
    expect(mainImg.src).toContain('src2');
    userEvent.click(leftArrow);
    expect(mainImg.src).toContain('src1');
  });

  it('should render change and display the right main img using the mini indicator imgs', () => {
    renderWithRouter(<ProductViewImgSlider imgs={imgs} />);
    const mainImg = screen.getByTestId('ZoomableImg') as HTMLImageElement;
    const miniImgSrc2 = screen.getByTestId('miniImg src2');
    const miniImgSrc3 = screen.getByTestId('miniImg src3');
    const miniImgSrc1 = screen.getByTestId('miniImg src1');

    expect(mainImg.src).toContain('src1');
    userEvent.click(miniImgSrc3);
    expect(mainImg.src).toContain('src3');
    userEvent.click(miniImgSrc1);
    expect(mainImg.src).toContain('src1');
    userEvent.click(miniImgSrc2);
    expect(mainImg.src).toContain('src2');
  });
});
