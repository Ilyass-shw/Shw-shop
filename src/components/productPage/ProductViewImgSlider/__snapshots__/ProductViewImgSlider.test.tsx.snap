// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductViewImgSlider should match Snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-jRQAMF kBeNDQ"
  >
    <div
      class="sc-dkPtyc bKvboz"
    >
      <button
        class="sc-gsDJrp sc-hKwCoD iyweFA kHQDkE"
        data-testid="leftArrow"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
          />
        </svg>
      </button>
      <div
        data-rmiz-wrap="visible"
      >
        <img
          class="sc-eCImvq iWIphU"
          data-testid="ZoomableImg"
          loading="lazy"
          src="src1"
        />
        <button
          aria-label="Zoom image"
          data-rmiz-btn-open="true"
          type="button"
        />
      </div>
      <button
        class="sc-gsDJrp sc-hKwCoD iyweFA itmmpb"
        data-testid="rightArrow"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
          />
        </svg>
      </button>
    </div>
    <div
      class="sc-gKckTs bOtfrW"
      data-testid="miniImgs"
    >
      <button
        class="sc-gsDJrp iyweFA"
      >
        <div
          class="sc-iCfLBT dHoujM"
        >
          <img
            class="sc-furvIG ccTpGe"
            data-testid="miniImg src1"
            loading="lazy"
            src="src1"
          />
        </div>
      </button>
      <button
        class="sc-gsDJrp iyweFA"
      >
        <div
          class="sc-iCfLBT ZbFQt"
        >
          <img
            class="sc-furvIG ccTpGe"
            data-testid="miniImg src2"
            loading="lazy"
            src="src2"
          />
        </div>
      </button>
      <button
        class="sc-gsDJrp iyweFA"
      >
        <div
          class="sc-iCfLBT ZbFQt"
        >
          <img
            class="sc-furvIG ccTpGe"
            data-testid="miniImg src3"
            loading="lazy"
            src="src3"
          />
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
